### PR TITLE
[vitaled] Brumberg vitaLED Binding initial contribution

### PR DESCRIPTION
**Pull request by MS1982**
**Initial development for Brumberg vitaLED binding**

This is the initial development for Brumberg vitaLED binding [vitaled]. This binding integrates Brumberg vitaLED via the Brumberg vitaLED LAN master that communicates via DMX with the Brumberg vitaLED bulbs. The LAN master supports 8 zones. Each zone can consists of 6 vitaLED bulbs. This binding supports the Brumberg vitaLED LAN master (Brumberg Article number 18308000) as thing.

The binding has in the current version the following limitations. Only none is supported as authentication yet. Do not use DHCP client, use instead a fix IP address. Groups are also not supported, but this can be handled via openHAB rule. Colour management and vital gradient are not yet supported.

The new development of the binding is "discussed" in this [thread](https://community.openhab.org/t/new-binding-for-brumberg-vitaled/39310) in the openhab community.

This is my first pull request. I followed the coding guidelines, the static code analysis and signed off my development. I'm looking forward to see my binding as part of openhab. :)

### CODE REVIEW

**Comment by martinvw** [comment]
You touched a lot of files which are not yours, I think its a good moment to get rid of those files and make sure you have squashed commit already:

git reset HEAD~17

Then commit again but only include files you are supposed to have changed and revert changes to all others, you need to force push to overwrite the changes which you pushed before. It could be wise to backup first, for example by creating a copy of this branch.

**Comment by MS1982** [comment]
Hi Martin, 

thanks for that feedback. I'm new to git and github and I asked myself if it's only possible to do a pull request for the complete addon branch. However, I have found no other way. I will try to clean up everything as you have described it.

BR Marcel

**Comment by MS1982** [comment]
I think now it's ok. Thanks for the help.

**Comment by martinvw** [comment]
I believe you should be able to link the marketplace to build output of this PR, see also the community forum.

Because now the travis build fails:

```
[ERROR] Code Analysis Tool has found: 
 1 error(s)! 
 1 warning(s) 
 0 info(s)
[ERROR] .binding.vitaled/build.properties:[0]
There is a jar outside of the lib folder /home/travis/build/openhab/openhab2-addons/addons/binding/org.openhab.binding.vitaled/org.openhab.binding.vitaled-2.3.0-SNAPSHOT.jar
[WARNING] org.openhab.binding.vitaled.handler.VitaLEDHandler.java:[177]
For periodically executed jobs that do not require a fixed rate scheduleWithFixedDelay should be preferred over scheduleAtFixedRate.
```

**Comment by MS1982** [comment]
Now it should be removed.

File: addons/binding/org.openhab.binding.vitaled/ESH-INF/thing/thing-types.xml [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+			<channel id="achromaticLight" typeId="achromaticLight">
+				<label>Achromatic Light</label>
+				<description>Achromatic Light in Kelvin</description>
+			</channel>
+			<channel id="intensity" typeId="intensity">
+				<label>Intensity</label>
+				<description>Intensity in %</description>
+			</channel>
+			<!-- Chromatic Light - RGBW control -->
+			<channel id="red" typeId="redLight">
```
Comment (triller-telekom) [inline addons/binding/org.openhab.binding.vitaled/ESH-INF/thing/thing-types.xml]: Do you only want to give the users direct access to the raw RGBW channels or would you also like to provide a `HSBType` channel, so openHAB users can use a `ColorPicker` to select a color?

**Comment by triller-telekom** [comment]
One more comment from my side: Since the vitaLED DMX lights seem to work with the [ESH DMX binding](https://github.com/eclipse/smarthome/tree/master/extensions/binding/org.eclipse.smarthome.binding.dmx), would it make sense to add you code as a new `bridge` to this binding instead of requiring the user to install a separate binding? So instead of buying a USB-to-DMX stick users could buy the vitaLED LAN master and use their their lights also via that one.

WDYT?

**Comment by triller-telekom** [comment]
@martinvw @kaikreuzer Can you please add an awaiting feedback label on this PR?

**Comment by MS1982** [comment]
Thanks for the review - sorry for my late reply - I'm very busy at the moment. Well, I'm not a Java developer therefore maybe I make things more complicated. :) However, I really appreciate this valuable feedback. I will adjust the code using the HttpUtil instead of HttpURLConnection. I also try to improve the exception handling, but as I said I'm not in that detail in Java. I also want to support HSB - so far as I understood I only have to convert HSB values to RGB values and vice versa.

Regarding the DMX bridge I'm not sure if this is a good idea in terms of finding the bridge for vitaLED. Before I have started developing this binding I have also asked in the openHab Forum if someone has a binding or knows a solution to control the vitaLED from openHabe. Then when I finished my initial development and published the binding I received the first reply that Laurens van Acker use OpenDMX with a USB stick to connect to vitaLED. I also don't know if the non-technical users understand that the vitaLED LAN master works as bridge to send HTTP Requests via LAN that are then send DMX commands to the vitaLED bulbs. I think it's probably easier to find and install a plugin with the name of the device instead of thinking how the device works and where you can maybe found a bridge for in another binding.

I will try to add additional COMMITS - I'm sorry I'm also new to github.

Best Regards,
Marcel

**Comment by davidgraeff** [comment]
Something went wrong here :/

**Comment by MS1982** [comment]
> 
> 
> Something went wrong here :/

Hm sorry, I don't know what happen - git will not become my friend... I have updated my binding and installing the current IDE. I wanted to update my github repository with the changes in my addon.

**Comment by MS1982** [comment]
How can I undo this?

**Comment by davidgraeff** [comment]
Easiest way if no have no idea about git is to checkout a fresh openhab2-addons and copy your binding dir into it (and update the pom.xml, features/karaf file).

Revert the last commit via the interactive rebase dialog: `git rebase -i HEAD~2` and follow the instructions.

**Comment by MS1982** [comment]
Ok thanks. I'm actually installing a new IDE and want to move my development to that in order to build a 2.5 snapshot.

**Comment by MS1982** [comment]
> 
> 
> Easiest way if you have no idea about git is to checkout a fresh openhab2-addons and copy your binding dir into it (and update the pom.xml, features/karaf file).
> 
> Revert the last commit via the interactive rebase dialog: `git rebase -i HEAD~2` and follow the instructions.

I have tried that, however I get an error that there are conflicts. However, I try to setup a new repository with 2,5 addon developement and then maybe I can pull that.

**Comment by MS1982** [comment]
> 
> 
> @MS1982: No need to hurry or rush things. We put these labels on PRs so we know where we are waiting for the binding developer and where people await feedback from the community.
> 
> Converting a `HSBType` from and to RGB is already build in, see:
> 
> https://github.com/eclipse/smarthome/blob/fd00823b2366ebcbc6fb32a69ea399f9f73a27bf/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/HSBType.java#L100
> 
> and
> 
> https://github.com/eclipse/smarthome/blob/fd00823b2366ebcbc6fb32a69ea399f9f73a27bf/bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/library/types/HSBType.java#L255
> 
> Regarding the discussion on whether to include the vitaLED support as a bridge to the existing DMX binding: I had another look at the channels that your binding supports and except for the colours they are different than the ones which are supported in the DMX binding.
> 
> Also the vitaLED lights work with the DMX binding, but can you use a generic DMX light with the LAN2DMX device from vitaLED? probably not because you cannot set the DMX id, etc., right? This sounds like a separate binding like yours makes sense.
> 
> Don't hesitate to ask if you have any questions regarding git or java!

I have now added a color channel in order to use the colorpicker. However, I have noticed that in the basic UI the selected color does not match from the result. On the iOS app it is much better.

However, I'm using the following method, but my bulbs have also a white LED at the moment I set the white color to 0. 

`PercentType[] rgb = hsbCommand.toRGB();`

Is there any method available for RGBW conversion?

**Comment by davidgraeff** [comment]
> Is there any method available for RGBW conversion?

Unfortunately not. I have asked Kai about that some time ago for the MiLight binding and he told me to use a second brightness channel for white

**Comment by J-N-K** [comment]
@MS1982, are you still working on this binding?

**Comment by Hilbrand** [comment]
@MS1982 Do you want to finish this binding? There are still a number of files in here that should not be in this pull request. The binding also needs to be updated to the latest build system changes. There is a tutorial https://community.openhab.org/t/tutorial-migrate-your-binding-to-the-maven-bnd-based-build-system/81389 to help migrating. That can also help cleaning up your branch here. Regarding the tutorial. Due to the changes for 3.0. development you should base your binding on the `2.5.x` branch instead of the `master` branch. If you need any help let us know. If you are not able to finish this binding can you close this pull request.

**Comment by Hilbrand** [comment]
Closing pr for long inactivity. If you still want to contribute please open a new pr.

**Pull request closed without merging**